deepchecks/vision/vision_data/vision_data.py

```diff
@@ -0,0 +1,239 @@
+        If True and the dynamic loader is of known type that can be shuffled, it will be shuffled.
+    """
+
+    def __init__(
+            self,
+            dynamic_loader,
+            task_type: str,
+            label_map: t.Optional[t.Dict[int, str]] = None,
+            dataset_name: t.Optional[str] = None,
+            reshuffle_dynamic_loader: bool = True
+    ):
+        self._dynamic_loader = shuffle_dynamic_loader(dynamic_loader) if reshuffle_dynamic_loader else dynamic_loader
+
+        if task_type not in ['classification', 'semantic_segmentation', 'object_detection', 'other']:
+            raise ValueError(f'Invalid task type: {task_type}, must be one of the following: '
+                             f'"classification", "semantic_segmentation", "object_detection", "other"')
+        self._task_type = TaskType(task_type)
+
+        if label_map is not None and not isinstance(label_map, dict):
+            raise ValueError('label_map must be a dictionary')
+        self._label_map = label_map
+        self.name = dataset_name
+
+        # indicator will be set to true in 'validate' method if the user implements the relevant formatters
+        self._has_images, self._has_labels, self._has_predictions = False, False, False
+        self._has_additional_data, self._has_embeddings, self._has_image_identifiers = False, False, False
+        self.validate()
+        self.init_cache()
+
+    def init_cache(self):
+        """Initialize the cache."""
+        self._num_images_seen = 0
+        # dict of class_id to number of images observed with label (num_label) and prediction (num_pred)
+        self._observed_classes = defaultdict()
+
+    def update_cache(self, batch_size, numpy_labels, numpy_predictions):
+        """update cache based on newly arrived batch."""
+        self._num_images_seen += batch_size
+        if numpy_labels is not None:
+            for class_id, num_observed in get_class_ids_from_numpy_labels(numpy_labels, self._task_type).items():
+                if class_id not in self._observed_classes:
+                    self._observed_classes[class_id] = {'num_label': 0, 'num_pred': 0}
+                self._observed_classes[class_id]['num_label'] += num_observed
+
+        if numpy_predictions is not None:
+            for class_id, num_observed in get_class_ids_from_numpy_preds(numpy_predictions, self._task_type).items():
+                if class_id not in self._observed_classes:
+                    self._observed_classes[class_id] = {'num_label': 0, 'num_pred': 0}
+                self._observed_classes[class_id]['num_pred'] += num_observed
+
+    def validate(self):
+        """Validate the VisionData functionalities implemented by the user and set which formatters were implemented."""
+        batch: BatchOutputFormat = next(iter(self._dynamic_loader))
+        allowed_keys = {'images', 'labels', 'predictions', 'additional_data', 'embeddings', 'image_identifiers'}
```

Comment:
Can't this validation be built in / use the BatchOutputFormat object?

Reply:
No, cause the user doesnt need to create a BatchOutputFormat object, he can also pass a simple dict